### ISSUE TITLE
Dynamic User Interface for a TOM

### ISSUE THREAD

**Issue by rachel3834** [TOMToolkit/tom_base#635]
**Is your feature request related to a problem? Please describe.**
TOM users would like to be able to edit, add and remove elements of the TOM UI without having to customise the code. 
Examples of desired flexibility include tables with built in search and sort functions, and the ability for a user to x-out columns they do not need.  
Another example would be dragging and dropping elements of a View to produce a layout more conducive to the user's workflow.  

**Describe the solution you'd like**
Modern UI libraries offer some of this functionality, so initially research use-cases in detail and then solutions to implement them.

**Comment by jchate6** [comment]
This would definitely involve a major departure from our current TOM Toolkit philosophy of avoiding javascript as much as possible.

I think an important point of clarification is understanding if we are allowing a TOM admin to dynamically adjust the UI that all users of a TOM would access, or if the goal is for each individual TOM user to be able to adjust and save the UI themselves. Basically, is the objective to help the TOM Admin, or the individual TOM users.

I think it's also important to know if there are specific features that are difficult for people to adjust to their needs through coding, or if the idea is to do more of a revamp to the entire frontend generally.

**Comment by rachel3834** [comment]
Yes, I accept that this is a significant departure.  We need to make sure
that TOM offers state-of-the-art UI/UX or the project risks creating a
negative impression for users.  At the same time, the goal of ensuring it
is easy to use and maintain is a core principle, so I'd like to explore
creative ways we can offer advanced functionality with the minimum of
overheads for the admin.

The purpose of this issue is really to capture a discussion thread about
this where the users can access it and chime in.  (Apologies is this is a
misuse of the issue system).

Your questions about how extensive this change should be are the heart of
the problem.  I can imagine adding a suite of Vue or React Components as
optional plugins for the Toolkit, allowing user to make use of them if they
need to.  This would keep the base TOM simple to use.   There is certainly
plenty of utility in that option.

I can also imagine a fully dynamic frontend, and how this could make the
Toolkit more approachable for users who don't have time to cross the Django
learning curve.  I can sketch out some wireframes as an example.   I don't
think the two options are necessarily mutually exclusive.

One question right now is: is there any way to implement a dynamic frontend
that is optional?
If it would have to be mandatory, how much of a maintenance headache is
this really, to have to include Javascript?   We have some community
testers who I'm sure would be happy to act as guinepigs for that.

**Comment by jchate6** [comment]
I think this is the perfect use of the issue system.

In general there is a bit of conflict between two levels of customization here, I think, that should be a focus of this discussion.
The more dynamic we make the frontend features, the more difficult it will likely become for an admin to modify them for use cases we don't accommodate or foresee. 
For example in the current set up, we provide a relatively simple photometry table. If a user wants to change it they can copy and edit it, or extend it with some new functionality, or move it, or remove it all together with a little html. If we make a dynamic table, they might be able to do certain things more easily, like displaying or hiding preset columns, but the process of changing the layout, or including entirely new columns we didn't provide becomes much more difficult and would rely heavily on languages and tool kits other than html. If we make a FULLY dynamic frontend that incorporates these features dynamically as well, that's an additional step of complication when it comes to altering them. 

I think an optional app that implemented some level of dynamism to the frontend is a really interesting idea. I think we'll need to think hard about what that looks like and probably write up a more formal proposal with explicit features based on what the discussion leads to here.

**Comment by rachel3834** [comment]
That's a very good point.  One key point of feedback that I have received
is that the Toolkit is seen as approachable for a scientist to edit and
customize; this is one of the key "selling points" for the package for the
userbase.  Swish javascript tools are great if they are exactly what you
need, but off-putting if they require a high learning curve to modify
them.

This is why having a very simple base TOM, plus an optional library of
advanced plugins is a good model.  If some of those plugins could include
entire dynamic frontends then that would be awesome.

One idea I've been toying with is offering a set of dynamic frontends that
are quite customized for different science use-cases, and could offer
advanced functionality because the expectation would be that the user
wouldn't need to modify them too much.
For example, we can design some "template TOM"s with dynamic frontends for
MMA, stellar astrophysics, and Solar System.

**Comment by jchate6** [comment]
[Tom_alerts_dash](https://github.com/TOMToolkit/tom_alerts_dash) might be worth looking at and reviving.

**Comment by jchate6** [comment]
[tom-toolkit-component-lib](https://github.com/TOMToolkit/tom-toolkit-component-lib) is also worth a look.

**Comment by jchate6** [comment]
- #987
